src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationWriter.cs

```diff
@@ -246,14 +254,17 @@ private XmlQualifiedName GetPrimitiveTypeName(Type type)
         [RequiresUnreferencedCode(XmlSerializer.TrimSerializationWarning)]
         protected void WriteTypedPrimitive(string? name, string? ns, object o, bool xsiType)
         {
-            string? value;
+            string? value = null;
             string type;
             string typeNs = XmlSchema.Namespace;
             bool writeRaw = true;
             bool writeDirect = false;
             Type t = o.GetType();
             bool wroteStartElement = false;
+            bool? tryFormatResult = null;
+            int charsWritten = -1;
 
+            char[] buffer = _primitivesBuffer ??= new char[128];
             switch (Type.GetTypeCode(t))
```

Comment:
Would `ArrayPool<char>.Shared.Rent(128);` and returning at the end be better? I think this class can be used by multiple threads simultaneously if you have more than 1 thread serializaing concurrently using the same serializer instance. This shared buffer will become corrupted in that case.

Reply:
@mconnew  You are right for synchronization problems and I followed your suggestion and used ArrayPool<char>.

Reply:
@mconnew Unfortunately this renting decreased performance even worse than before. Here are samples using benchmark from here : https://github.com/dotnet/performance/pull/2623

Before:
|        Method |     Mean |     Error |    StdDev |   Median |      Min |      Max |   Gen0 | Allocated |
|-------------- |---------:|----------:|----------:|---------:|---------:|---------:|-------:|----------:|
| AddPrimitives | 1.878 us | 0.0175 us | 0.0163 us | 1.876 us | 1.857 us | 1.908 us | 0.0900 |     792 B |

After:
|        Method |     Mean |     Error |    StdDev |   Median |      Min |      Max |   Gen0 | Allocated |
|-------------- |---------:|----------:|----------:|---------:|---------:|---------:|-------:|----------:|
| AddPrimitives | 2.192 μs | 0.0433 μs | 0.0384 μs | 2.177 μs | 2.155 μs | 2.275 μs | 0.0400 |     360 B |

There is definitely win in memory, but regression in execution.
I've looked at where XmlSerializationWriter is instantiated and found only here :https://github.com/dotnet/runtime/blob/9b76c28567640e4cbe0d20e18b765b8f1a47473f/src/libraries/System.Private.Xml/src/System/Xml/Serialization/XmlSerializer.cs#L364.
Looks like it always creates new instance and there is method "Init" which also implies that this class have state and cannot be used concurrently. 
Do you remember where this class is used concurrently ?

Reply:
When using a pre-generated serializer, it provides an instance derived from XmlSerializationWriter which I think is cached and reused with multiple calls to Serialize. There's also the reflection based serialization used when it's using SOAP mapping which might not use a new instance each time. The allocation is small, what's the cost of allocating it fresh each time? You could also implement it in such a way that it's optimal for single usage. Eg have a `byte[]` field and use Interlocked.Exchange replacing it with null to "rent" the buffer. If another thread is using it, you will get a null back from the exchange so you create a new array locally. When you are done with it, use Interlocked.Exchange to put it back. This avoids the overhead of renting, will have zero allocation in the common case, and degrades to something no worse than it was originally.

Reply:
@mconnew I like your idea and pushed new commit with it. Could you check again ?
Also results now are better then before:
|        Method |     Mean |     Error |    StdDev |   Median |      Min |      Max |   Gen0 | Allocated |
|-------------- |---------:|----------:|----------:|---------:|---------:|---------:|-------:|----------:|
| AddPrimitives | 1.969 us | 0.0351 us | 0.0293 us | 1.962 us | 1.942 us | 2.047 us | 0.0400 |     360 B |

Reply:
@mconnew One more question. I create char[128] which might be too big. Any suggestion here? When I execute tests from Performance repo I see increase of bytes allocated(maximum 8), but I really hoped to see some reduction :(. 
Should I increase code complexity by starting with lower buffer like char[64] and then to inspect TryFormat methods and retry with bigger buffer, or we can leave this regression for small number of primitives, and have bigger wins when xml produces is much bigger ?
Here are results from XML performance comparison tests:

## Statistics

Total:   68
Same:    75.00 %
Slower:  5.88 %
Faster:  10.29 %
Noise:   8.82 %
Unknown: 0.00 %

## Statistics per Architecture

| Architecture | Same    | Slower | Faster  | Noise  | Unknown|
| ------------ | ------- | ------ | ------- | ------ | ------- |
| X64          | 75.00 % | 5.88 % | 10.29 % | 8.82 % | 0.00 % |

## Statistics per Operating System

| Operating System | Same    | Slower | Faster  | Noise  | Unknown|
| ---------------- | ------- | ------ | ------- | ------ | ------- |
| Windows 10       | 75.00 % | 5.88 % | 10.29 % | 8.82 % | 0.00 % |

## Statistics per Namespace

| Namespace                              | Same     | Slower  | Faster  | Noise   | Unknown|
| -------------------------------------- | -------- | ------- | ------- | ------- | ------- |
| MicroBenchmarks.Serializers            | 78.12 %  | 9.38 %  | 12.50 % | 0.00 %  | 0.00 % |
| Microsoft.Extensions.Configuration.Xml | 75.00 %  | 0.00 %  | 25.00 % | 0.00 %  | 0.00 % |
| System.Xml.Linq                        | 73.68 %  | 0.00 %  | 0.00 %  | 26.32 % | 0.00 % |
| System.Xml.Tests                       | 66.67 %  | 16.67 % | 16.67 % | 0.00 %  | 0.00 % |
| XmlDocumentTests.XmlDocumentTests      | 66.67 %  | 0.00 %  | 33.33 % | 0.00 %  | 0.00 % |
| XmlDocumentTests.XmlNodeListTests      | 100.00 % | 0.00 %  | 0.00 %  | 0.00 %  | 0.00 % |
| XmlDocumentTests.XmlNodeTests          | 50.00 %  | 0.00 %  | 0.00 %  | 50.00 % | 0.00 % |

Reply:
I have thought of a way to eliminate the allocation. You could use `stackalloc` to allocate the memory on the stack. The only problem with this is that XmlWriter needs a `char[]` passed which means you can't assign the stack allocated memory to a `Span<char>` initially, we need it as an array. marking the method as unsafe. `WriteTypedPrimitive` is a protected method on an existing public type. I don't know if adding the unsafe keyword is a breaking change. This is easily solved by moving the implementation to an inner private method which is marked unsafe and calling that from this method. So something like this:
```c#
protected void WriteTypedPrimitive(string? name, string? ns, object o, bool xsiType)
{
    UnsafeWriteTypedPrimitive(name, ns, o, xsiType);
}

private void UnsafeWriteTypedPrimitive(string? name, string? ns, object o, bool xsiType)
{
    char* buffer = stackalloc char[128];
    Span<char> bufferSpan = buffer;
    // rest of implementation
}
```

Reply:
@mconnew Unfortunately I am not so good with pointer math. 
How can I convert char* to char[]?
I cannot see any method of XmlWriter which accepts char*.
In XmlUtf8RawTextWriter I can see that there is iteration using char*, but that's all.

Reply:
I've been thinking about this some more and realized multiple methods can't be called on different threads concurrently.  The class wraps the XmlWriter (stored in the _w field) which means the thread usage must match that of XmlWriter. And as you can only have one thread at a time trying to write to an XmlWriter, only one thread at a time will be calling methods on this class. It's safe to have a single instance of the buffer which exists for the lifetime of this class.  
I tried to work out a way to cast a char* to a char[] and there's no good clean way to do it.